lib/autoupdate/agent/updater.go

```diff
@@ -0,0 +1,338 @@
+
+// UpdateSpec describes the spec field in update.yaml.
+type UpdateSpec struct {
+	// Proxy address
+	Proxy string `yaml:"proxy"`
+	// Group update identifier
+	Group string `yaml:"group"`
+	// URLTemplate for the Teleport tgz download URL.
+	URLTemplate string `yaml:"url_template"`
+	// Enabled controls whether auto-updates are enabled.
+	Enabled bool `yaml:"enabled"`
+}
+
+// UpdateStatus describes the status field in update.yaml.
+type UpdateStatus struct {
+	// ActiveVersion is the currently active Teleport version.
+	ActiveVersion string `yaml:"active_version"`
+}
+
+// NewLocalUpdater returns a new Updater that auto-updates local
+// installations of the Teleport agent.
+// The AutoUpdater uses an HTTP client with sane defaults for downloads, and
+// will not fill disk to within 10 MB of available capacity.
+func NewLocalUpdater(cfg LocalUpdaterConfig) (*Updater, error) {
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tr, err := libdefaults.Transport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tr.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: cfg.InsecureSkipVerify,
+		RootCAs:            certPool,
+	}
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   cfg.DownloadTimeout,
+	}
+	if cfg.Log == nil {
+		cfg.Log = slog.Default()
+	}
+	return &Updater{
+		Log:                cfg.Log,
+		Pool:               certPool,
+		InsecureSkipVerify: cfg.InsecureSkipVerify,
+		ConfigPath:         filepath.Join(cfg.VersionsDir, updateConfigName),
+		Installer: &LocalInstaller{
+			InstallDir: cfg.VersionsDir,
+			HTTP:       client,
+			Log:        cfg.Log,
+
+			ReservedFreeTmpDisk:     reservedFreeDisk,
+			ReservedFreeInstallDisk: reservedFreeDisk,
+		},
+	}, nil
+}
+
+// LocalUpdaterConfig specifies configuration for managing local agent auto-updates.
+type LocalUpdaterConfig struct {
+	// Log contains a slog logger.
+	// Defaults to slog.Default() if nil.
+	Log *slog.Logger
+	// InsecureSkipVerify turns off TLS certificate verification.
+	InsecureSkipVerify bool
+	// DownloadTimeout is a timeout for file download requests.
+	// Defaults to no timeout.
+	DownloadTimeout time.Duration
+	// VersionsDir for installing Teleport (usually /var/lib/teleport/versions).
+	VersionsDir string
+}
+
+// Updater implements the agent-local logic for Teleport agent auto-updates.
+type Updater struct {
+	// Log contains a logger.
+	Log *slog.Logger
+	// Pool used for requests to the Teleport web API.
+	Pool *x509.CertPool
+	// InsecureSkipVerify skips TLS verification.
+	InsecureSkipVerify bool
+	// ConfigPath contains the path to the agent auto-updates configuration.
+	ConfigPath string
+	// Installer manages installations of the Teleport agent.
+	Installer Installer
+}
+
+// Installer provides an API for installing Teleport agents.
+type Installer interface {
+	// Install the Teleport agent at version from the download template.
+	Install(ctx context.Context, version, template string) error
```

Comment:
The download URL template is also persisted into `update.yaml`, in the case that it is overridden by the user. It could be confusing to pass the template to `Enable` for storage, but not use `Enable`'s copy of the value during the installation that `Enable` initiates.